### PR TITLE
ImageSet now supports multiple sets

### PR DESCRIPTION
**Pull request by Upabjojr**
Additionally, `imageset` has been remapped to use multipledispatching.

### CODE REVIEW

File: sympy/sets/dispatchers.py [hunk 1]
```diff
@@ -150,6 +241,18 @@ def function_sets(f, x):
...
     return ImageSet(Lambda(_x, f(_x)), x)
 
-@dispatch(FunctionClass, Set)
+@dispatch(FunctionUnion, Union)
+def function_sets(f, x):
+    return Union(imageset(f, arg) for arg in x.args)
+
+@dispatch(FunctionUnion, Intersection)
+def function_sets(f, x):
+    return Intersection(imageset(f, arg) for arg in x.args)
```
Comment (jksuom) [inline sympy/sets/dispatchers.py]: Intersection of image sets may contain elements that are not in the image of the intersection. Compare the intersection of  `f([-1, 0])` and `f([0, 1])` with the image of the intersection `f({0})` when `f(x) = x**2`.

Comment (Upabjojr) [reply sympy/sets/dispatchers.py]: I see. This code was just copied from the `_eval_imageset` method of Intersection. It is supposed to be called when the intersection cannot be calculated.

Shall we leave the imageset unevaluated for this case?

Comment (Upabjojr) [reply sympy/sets/dispatchers.py]: Bijective functions may be treated separately.

Comment (jksuom) [reply sympy/sets/dispatchers.py]: > Shall we leave the imageset unevaluated for this case?

I guess that is what has to be done.

File: sympy/sets/dispatchers.py [hunk 2]
```diff
@@ -150,6 +241,27 @@ def function_sets(f, x):
...
+
+@dispatch(FunctionUnion, Intersection)
+def function_sets(f, x):
+    # If the function is invertible, intersect the maps of the sets.
+    u = symbols("u")
+    fdiff = f(u).diff(u)
+    # TODO: find a better condition for invertible functions:
+    if ((f in (exp, log))  # functions known to be invertible
+        or (fdiff > 0) == True or (fdiff < 0) == True  # monotonous funcs
+        ):
```
Comment (Upabjojr) [inline sympy/sets/dispatchers.py]: @jksuom I put these conditions of invertibility. `solveset` needs this mapping to solve equations with `Abs` and similar. It does not look that clean.

Do we have a better way to test the invertibility of functions?

Comment (jksuom) [reply sympy/sets/dispatchers.py]: We probably should have a method for testing invertibility in a set such as the union of `x.args`.

**Pull request merged**